src/main.rs

```diff
@@ -0,0 +1,905 @@
+) -> Result<DatasetEvent, String> {
+    match decoder.decode(msg.payload()) {
+        Ok(result) => match result.name {
+            Some(name) => match name.name.as_str() {
+                "DatasetEvent" => match name.namespace {
+                    Some(namespace) => match namespace.as_str() {
+                        "no.fdk.dataset" => match from_value::<DatasetEvent>(&result.value) {
+                            Ok(event) => Ok(event),
+                            Err(e) => Err(format!("Deserialization failed {}", e)),
+                        },
+                        ns => Err(format!("Unexpected namespace {}", ns)),
+                    },
+                    None => Err("No namespace in schema, while expected".to_string()),
+                },
+                name => Err(format!("Unexpected name {}", name)),
+            },
+            None => Err("No name in schema, while expected".to_string()),
+        },
+        Err(e) => Err(format!("error getting dataset-event: {}", e)),
+    }
+}
+
+// Read DatasetEvent message of type DATASET_HARVESTED
+fn handle_dataset_event<'a>(
+    msg: OwnedMessage,
+    decoder: AvroDecoder,
+) -> Result<MQAEvent, String> {
+    info!("Handle DatasetEvent on message {}", msg.offset());
+
+    let dataset_event = parse_dataset_event(msg, decoder);
+
+    match dataset_event {
+        Ok(event) => {
+            let dt = Utc.timestamp_millis(event.timestamp);
+            info!(
+                "{} - Processing dataset event with timestamp {:?}",
+                event.fdkId, dt
+            );
+            return parse_rdf_graph_and_calculate_metrics(event.fdkId, event.graph);
+        }
+        Err(e) => Err(format!("Unable to decode dataset event: {}", e)),
+    }
+}
+
+fn parse_rdf_graph_and_calculate_metrics(
+    fdk_id: String,
+    graph: String,
+) -> Result<MQAEvent, String> {
+    let store = parse_turtle(graph);
+
+    match get_dataset_node(&store) {
+        Some(dataset_node) => {
+            let metrics_store = calculate_metrics(dataset_node.as_ref(), &store);
+
+            match dump_graph_as_turtle(&metrics_store) {
+                Ok(bytes) => {
+                    // Create MQA event
+                    match str::from_utf8(bytes.as_slice()) {
+                        Ok(turtle) => Ok(MQAEvent {
+                            r#type: MQAEventType::PROPERTIES_CHECKED,
+                            fdkId: fdk_id,
+                            graph: turtle.to_string(),
+                            timestamp: Utc::now().timestamp_millis(),
+                        }),
+                        Err(e) => Err(format!("Failed dumping graph as turtle: {}", e)),
+                    }
+                }
+                Err(e) => Err(format!("Failed dumping graph as turtle: {}", e)),
+            }
+        }
+        None => Err(format!("{} - Dataset node not found in graph", fdk_id)),
+    }
+}
+
+fn calculate_metrics(dataset_node: NamedNodeRef, store: &Store) -> Store {
```

Comment:
Innenfor denne funksjonen er det en del muligheter for panic dersom noe feiler. Alle feilene ser ut til å være av typen `StorageError`. Denne kunne i stedet vært returnert oppover ved bruk av `Result`. Hva tror du om at `calculate_metrics`, `create_metrics_store` og `add_quality_measurement` returnerer `Result<, StorageError>`. Da kan en del `.unwrap()` erstattes med `?`.